File: src/chain-observers/providers/witness-provider.ts
import { TAccountName } from "@hiveio/wax";
import { TRegisterEvaluationContext } from "../classifiers/collector-classifier-base";
import { IWitness, WitnessClassifier } from "../classifiers/witness-classifier";
import { TProviderEvaluationContext } from "../factories/data-evaluation-context";
import { ProviderBase } from "./provider-base";

export type TWitnessProvider<TAccounts extends Array<TAccountName>> = {
  [K in TAccounts[number]]: IWitness;
};

export interface IWitnessProviderData<TAccounts extends Array<TAccountName>> {
  witnesses: Partial<TWitnessProvider<TAccounts>>;
};

export interface IWitnessProviderOptions {
  accounts: string[];
}

export class WitnessProvider<
  TAccounts extends Array<TAccountName> = Array<TAccountName>
> extends ProviderBase<IWitnessProviderOptions, IWitnessProviderData<TAccounts>> {
  public readonly witnesses = new Set<TAccountName>();

  public pushOptions(options: IWitnessProviderOptions): void {
    for(const account of options.accounts)
      this.witnesses.add(account);
  }

  public usedContexts(): Array<TRegisterEvaluationContext> {
    const classifiers: Array<TRegisterEvaluationContext> = [];
    for(const witness of this.witnesses)
      classifiers.push(WitnessClassifier.forOptions({ witness }));

    return classifiers;
  }

  public get baseStructure(): IWitnessProviderData<TAccounts> {
    return {
      witnesses: {}
    };
  }

  public async provide(data: TProviderEvaluationContext): Promise<IWitnessProviderData<TAccounts>> {
    const result = this.baseStructure;

    const { witnesses } = await data.get(WitnessClassifier);
    for(const witness of this.witnesses)
      result.witnesses[witness] = witnesses[witness];

    return result as IWitnessProviderData<TAccounts>;
  }
}

File: python/workerbee/chain_observers/providers/witness_provider.py
"""WitnessProvider — provides witness data for tracked accounts.

Mirrors src/chain-observers/providers/witness-provider.ts.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from .._ordered_set import OrderedSet
from ..classifiers.collector_classifier_base import TRegisterEvaluationContext
from ..classifiers.witness_classifier import WitnessClassifier
from .provider_base import ProviderBase

if TYPE_CHECKING:
    from ..factories.data_evaluation_context import DataEvaluationContext
    from ..payloads import WitnessData, WitnessesPayload


class WitnessProvider(ProviderBase):
    def __init__(self) -> None:
        self.witnesses: OrderedSet[str] = OrderedSet()

    def push_options(self, options: Any) -> None:
        for account in options["accounts"]:
            self.witnesses.add(account)

    def used_contexts(self) -> list[TRegisterEvaluationContext]:
        classifiers: list[TRegisterEvaluationContext] = []
        for witness in self.witnesses:
            classifiers.append(
                WitnessClassifier.for_options({"witness": witness}),
            )
        return classifiers

    async def provide(self, data: DataEvaluationContext) -> WitnessesPayload:
        witness_data = await data.get(WitnessClassifier)
        source = witness_data["witnesses"]

        witnesses: dict[str, WitnessData | None] = {}
        for witness in self.witnesses:
            witnesses[witness] = source.get(witness)

        return {"witnesses": witnesses}
